extensions/amp-viewer-integration/0.1/messaging/messaging.js

```diff
@@ -116,15 +116,85 @@ export class WindowPortEmulator {
  * ampdoc and Bob is the viewer.
  */
 export class Messaging {
+  /**
+   * Performs a handshake and initializes messaging.
+   * @param {!Window} source
+   * @param {!Window} target
+   * @param {?string=} opt_token
+   * @param {?number=} opt_interval
+   * @return {!Promise<!Messaging>}
+   */
+  static initiateHandshakeWithDocument(
+    source,
+    target,
+    opt_token,
+    opt_interval
+  ) {
+    const interval = opt_interval || 1000;
+    return new Promise(resolve => {
+      const intervalRef = setInterval(() => {
+        const channel = new MessageChannel();
+        const pollMessage = /** @type {JsonObject} */ ({
+          app: APP,
+          name: HANDSHAKE_POLL_MSG,
+        });
+        target./*OK*/ postMessage(pollMessage, '*', [channel.port2]);
+
+        const port = channel.port1;
+        const listener = e => {
+          const data = {e};
+          if (data['app'] === APP && data['name'] === CHANNEL_OPEN_MSG) {
+            clearInterval(intervalRef);
+            port.removeEventListener('message', listener);
+            const messaging = new Messaging(null, port, false, opt_token, true);
+            messaging.sendResponse_(data['requestid'], CHANNEL_OPEN_MSG, null);
+            resolve(messaging);
+          }
+        };
+        port.addEventListener('message', listener);
+        port.start();
+      }, interval);
+    });
+  }
+
+  /**
+   * Waits for handshake from iframe and initializes messaging.
```

Comment:
Similarly, let's document the requirement for `origin` viewer parameter.

Reply:
Sorry, not sure I understand: what viewer parameter is needed for `origin`?

Reply:
I believe amp-viewer-integration.js requires it in this mode. Is that not what you're seeing?

Reply:
Checked the code and you're absolutely right. It won't initiate a handshake if no `origin` is specified. I'm also wrong about `handshakepoll`, that one is required for the other method, not for this one. Fixed.